// @wdio/testingbot-service/build/service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const got_1 = __importDefault(require("got"));
const logger_1 = __importDefault(require("@wdio/logger"));
const log = logger_1.default('@wdio/testingbot-service');
const jobDataProperties = ['name', 'tags', 'public', 'build', 'extra'];
class TestingBotService {
    constructor() {
        this._failures = 0;
        this._testCnt = 0;
    }
    beforeSession(config, capabilities) {
        this._config = config;
        this._capabilities = capabilities;
        this._config.user = config.user;
        this._config.key = config.key;
        this._tbUser = this._config.user;
        this._tbSecret = this._config.key;
        this._isServiceEnabled = Boolean(this._tbUser && this._tbSecret);
    }
    before(caps, specs, browser) {
        this._browser = browser;
    }
    beforeSuite(suite) {
        this._suiteTitle = suite.title;
    }
    beforeTest(test) {
        if (!this._isServiceEnabled || !this._browser) {
            return;
        }
        if (this._suiteTitle === 'Jasmine__TopLevel__Suite') {
            this._suiteTitle = test.fullName.slice(0, test.fullName.indexOf(test.title) - 1);
        }
        const context = (test.fullName ||
            `${test.parent} - ${test.title}`);
        this._browser.execute('tb:test-context=' + context);
    }
    afterSuite(suite) {
        if (Object.prototype.hasOwnProperty.call(suite, 'error')) {
            ++this._failures;
        }
    }
    afterTest(test, context, results) {
        if (!results.passed) {
            ++this._failures;
        }
    }
    beforeFeature(uri, feature) {
        if (!this._isServiceEnabled || !this._browser) {
            return;
        }
        this._suiteTitle = feature.document.feature.name;
        this._browser.execute('tb:test-context=Feature: ' + this._suiteTitle);
    }
    beforeScenario(uri, feature, scenario) {
        if (!this._isServiceEnabled || !this._browser) {
            return;
        }
        const scenarioName = scenario.name;
        this._browser.execute('tb:test-context=Scenario: ' + scenarioName);
    }
    afterScenario(uri, feature, pickle, result) {
        if (result.status === 'failed') {
            ++this._failures;
        }
    }
    after(result) {
        var _a;
        if (!this._isServiceEnabled || !this._browser) {
            return;
        }
        let failures = this._failures;
        if (((_a = this._browser.config.mochaOpts) === null || _a === void 0 ? void 0 : _a.bail) && Boolean(result)) {
            failures = 1;
        }
        const status = 'status: ' + (failures > 0 ? 'failing' : 'passing');
        if (!this._browser.isMultiremote) {
            log.info(`Update job with sessionId ${this._browser.sessionId}, ${status}`);
            return this.updateJob(this._browser.sessionId, failures);
        }
        const browser = this._browser;
        return Promise.all(Object.keys(this._capabilities).map((browserName) => {
            log.info(`Update multiremote job for browser "${browserName}" and sessionId ${browser[browserName].sessionId}, ${status}`);
            return this.updateJob(browser[browserName].sessionId, failures, false, browserName);
        }));
    }
    onReload(oldSessionId, newSessionId) {
        if (!this._isServiceEnabled || !this._browser) {
            return;
        }
        const status = 'status: ' + (this._failures > 0 ? 'failing' : 'passing');
        if (!this._browser.isMultiremote) {
            log.info(`Update (reloaded) job with sessionId ${oldSessionId}, ${status}`);
            return this.updateJob(oldSessionId, this._failures, true);
        }
        const browser = this._browser;
        const browserName = browser.instances.filter((browserName) => browser[browserName].sessionId === newSessionId)[0];
        log.info(`Update (reloaded) multiremote job for browser "${browserName}" and sessionId ${oldSessionId}, ${status}`);
        return this.updateJob(oldSessionId, this._failures, true, browserName);
    }
    async updateJob(sessionId, failures, calledOnReload = false, browserName) {
        if (!this._browser) {
            return;
        }
        const json = this.getBody(failures, calledOnReload, browserName);
        this._failures = 0;
        const response = await got_1.default.put(this.getRestUrl(sessionId), {
            json,
            responseType: 'json',
            username: this._tbUser,
            password: this._tbSecret
        });
        this._browser.jobData = response.body;
        return response.body;
    }
    getRestUrl(sessionId) {
        return `https://api.testingbot.com/v1/tests/${sessionId}`;
    }
    getBody(failures, calledOnReload = false, browserName) {
        let body = { test: {} };
        body.test['name'] = this._suiteTitle;
        if ((calledOnReload || this._testCnt) && this._browser) {
            let testCnt = ++this._testCnt;
            if (this._browser.isMultiremote) {
                testCnt = Math.ceil(testCnt / this._browser.instances.length);
            }
            body.test['name'] += ` (${testCnt})`;
        }
        for (let prop of jobDataProperties) {
            if (!this._capabilities[prop]) {
                continue;
            }
            body.test[prop] = this._capabilities[prop];
        }
        if (browserName) {
            body.test['name'] = `${browserName}: ${body.test['name']}`;
        }
        body.test['success'] = failures === 0 ? '1' : '0';
        return body;
    }
}
exports.default = TestingBotService;
